head/src/main/kotlin/io/qalipsis/core/head/report/ReportFileBuilder.kt

```diff
@@ -0,0 +1,127 @@
+     * @param tenant reference of the tenant owning the report
+     * @param campaignData data class which contains all report data
+     */
+    suspend fun execute(
+        report: ReportEntity,
+        tenant: String,
+        campaignData: Collection<CampaignData>
+    ): CampaignReportDetail {
+        campaignData.forEach { campaign ->
+            val zones = mutableSetOf<Zone>()
+            headConfiguration.zones.forEach { zone ->
+                if (zone.key in campaign.zones) {
+                    zones.add(zone)
+                }
+            }
+            campaign.apply { resolvedZones = zones }
+        }
+        val tableData = mutableListOf<Collection<TimeSeriesRecord>?>()
+        val chartData = mutableListOf<Map<String, List<TimeSeriesAggregationResult>>>()
+        report.dataComponents.forEach { fetchDataByComponentType(tenant, report, it, tableData, chartData) }
+
+        return CampaignReportDetail(
+            reportName = report.displayName,
+            campaignData = campaignData,
+            tableData = tableData,
+            chartData = chartData
+        )
+    }
+
+    /**
+     *  Fetches the needed data for report charts or report tables.
+     *
+     *  @param tenant reference of the tenant owning the campaign
+     *  @param report current [ReportEntity] to be displayed
+     *  @param reportDataComponentEntity specifies the data component type
+     *  @param tableData holds a collection of information needed to populate report tables
+     *  @param chartData holds a collection of information needed to populate report chart
+     */
+    @KTestable
+    private suspend fun fetchDataByComponentType(
+        tenant: String,
+        report: ReportEntity,
+        reportDataComponentEntity: ReportDataComponentEntity,
+        tableData: MutableCollection<Collection<TimeSeriesRecord>?>,
+        chartData: MutableCollection<Map<String, List<TimeSeriesAggregationResult>>>
+    ) {
+        val campaignReferences = report.campaignKeys.ifEmpty { report.campaignNamesPatterns }
+        val dataSeries = reportDataComponentEntity.dataSeries.map { it.reference }.toSet()
+        when (reportDataComponentEntity.type) {
+            DataComponentType.DATA_TABLE -> tableData.add(
+                fetchTableData(
+                    tenant,
+                    dataSeries,
+                    campaignReferences,
+                    report
+                )
+            )
+
+            DataComponentType.DIAGRAM -> chartData.add(fetchChartData(tenant, campaignReferences, report, dataSeries))
+        }
+    }
+
+    /**
+     * Fetches data from [TimeSeriesDataQueryService] to be used in populating report tables.
+     */
+    @KTestable
+    private suspend fun fetchTableData(
+        tenant: String,
+        dataSeries: Set<String>,
+        campaignReferences: Collection<String>,
+        report: ReportEntity
+    ): Set<TimeSeriesRecord> {
```

Comment:
Do not do a Set here, the order matters. You have to keep the data in the order they are received.
```suggestion
    ): List<TimeSeriesRecord> {
```